File: mobile/android/EventNexusScanner/app/src/main/java/eu/eventnexus/scanner/viewmodel/ScannerViewModel.kt
package eu.eventnexus.scanner.viewmodel

import android.app.Application
import android.content.Context
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import eu.eventnexus.scanner.data.EventInfo
import eu.eventnexus.scanner.data.ScanResult
import eu.eventnexus.scanner.data.TicketInfo
import eu.eventnexus.scanner.network.SupabaseApi
import eu.eventnexus.scanner.network.SupabaseClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.*

/**
 * ViewModel for scanner functionality
 * Manages authentication state, scanning state, and API communication
 */
class ScannerViewModel(application: Application) : AndroidViewModel(application) {
    
    private val api: SupabaseApi = SupabaseClient.api
    private val prefs = application.getSharedPreferences("scanner_prefs", Context.MODE_PRIVATE)
    
    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()
    
    private val _currentEvent = MutableStateFlow<EventInfo?>(null)
    val currentEvent: StateFlow<EventInfo?> = _currentEvent.asStateFlow()
    
    private val _scannerCodeId = MutableStateFlow<String?>(null)
    val scannerCodeId: StateFlow<String?> = _scannerCodeId.asStateFlow()
    
    private val _isScanning = MutableStateFlow(false)
    val isScanning: StateFlow<Boolean> = _isScanning.asStateFlow()
    
    private val _scannedCount = MutableStateFlow(0)
    val scannedCount: StateFlow<Int> = _scannedCount.asStateFlow()
    
    private val _isConnected = MutableStateFlow(true)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()
    
    private val _sessionStartTime = MutableStateFlow<Long?>(null)
    private val _sessionDuration = MutableStateFlow("00:00")
    val sessionDuration: StateFlow<String> = _sessionDuration.asStateFlow()
    
    private var scannerCode: String? = null
    
    init {
        // Restore saved session if exists
        restoreSession()
        
        // Update session duration every second
        viewModelScope.launch {
            while (true) {
                _sessionStartTime.value?.let { startTime ->
                    val duration = (System.currentTimeMillis() - startTime) / 1000
                    val hours = duration / 3600
                    val minutes = (duration % 3600) / 60
                    val seconds = duration % 60
                    _sessionDuration.value = if (hours > 0) {
                        String.format("%d:%02d:%02d", hours, minutes, seconds)
                    } else {
                        String.format("%02d:%02d", minutes, seconds)
                    }
                }
                kotlinx.coroutines.delay(1000)
            }
        }
    }
    
    /**
     * Restore saved session from SharedPreferences
     */
    private fun restoreSession() {
        val savedCode = prefs.getString("scanner_code", null)
        val savedCodeId = prefs.getString("scanner_code_id", null)
        val savedEventId = prefs.getString("event_id", null)
        val savedEventName = prefs.getString("event_name", null)
        val savedEventDate = prefs.getString("event_date", null)
        val savedEventLocation = prefs.getString("event_location", null)
        val savedSessionStart = prefs.getLong("session_start", 0)
        val savedScannedCount = prefs.getInt("scanned_count", 0)
        
        if (!savedCode.isNullOrEmpty() && !savedCodeId.isNullOrEmpty()) {
            scannerCode = savedCode
            _scannerCodeId.value = savedCodeId
            _scannedCount.value = savedScannedCount
            
            if (savedSessionStart > 0) {
                _sessionStartTime.value = savedSessionStart
            }
            
            if (!savedEventId.isNullOrEmpty() && !savedEventName.isNullOrEmpty()) {
                _currentEvent.value = EventInfo(
                    id = savedEventId,
                    name = savedEventName,
                    date = savedEventDate,
                    location = savedEventLocation
                )
            }
            
            _isAuthenticated.value = true
        }
    }
    
    /**
     * Save session to SharedPreferences
     */
    private fun saveSession() {
        prefs.edit().apply {
            putString("scanner_code", scannerCode)
            putString("scanner_code_id", _scannerCodeId.value)
            putString("event_id", _currentEvent.value?.id)
            putString("event_name", _currentEvent.value?.name)
            putString("event_date", _currentEvent.value?.date)
            putString("event_location", _currentEvent.value?.location)
            putLong("session_start", _sessionStartTime.value ?: 0)
            putInt("scanned_count", _scannedCount.value)
            apply()
        }
    }
    
    /**
     * Clear saved session
     */
    private fun clearSession() {
        prefs.edit().clear().apply()
    }
    
    /**
     * Authenticate with scanner code
     */
    suspend fun authenticateWithCode(code: String) {
        try {
            println("🔍 Calling API with code: $code")
            val response = api.verifyScannerCode(mapOf("p_code" to code))
            println("📡 API Response: $response")
            
            if (response.isNotEmpty()) {
                val result = response[0]
                println("📊 Result valid: ${result.valid}, event_id: ${result.event_id}, scanner_code_id: ${result.scanner_code_id}")
                
                if (result.valid) {
                    scannerCode = code
                    _scannerCodeId.value = result.scanner_code_id
                    _sessionStartTime.value = System.currentTimeMillis()
                    
                    // Fetch event details
                    result.event_id?.let { eventId ->
                        println("🎉 Fetching event details for: $eventId")
                        fetchEventDetails(eventId)
                    }
                    
                    _isAuthenticated.value = true
                    println("✅ Authentication successful, saving session...")
                    
                    // Save session to SharedPreferences
                    saveSession()
                    
                    // Record usage
                    result.scanner_code_id?.let { 
                        println("📝 Recording usage for scanner code: $it")
                        recordUsage(it) 
                    }
                } else {
                    println("❌ Scanner code invalid or expired")
                    throw Exception("Invalid scanner code or code expired")
                }
            } else {
                println("❌ Empty response from API")
                throw Exception("Invalid scanner code")
            }
        } catch (e: retrofit2.HttpException) {
            println("🌐 HTTP Error: ${e.code()} - ${e.message()}")
            throw Exception("Failed to connect to server: ${e.message()}")
        } catch (e: java.net.UnknownHostException) {
            println("🌐 Network Error: No internet connection")
            throw Exception("Failed to connect: Check your internet connection")
        } catch (e: Exception) {
            println("💥 Exception: ${e.javaClass.simpleName} - ${e.message}")
            e.printStackTrace()
            throw Exception("Authentication failed: ${e.message}")
        }
    }
    
    /**
     * Fetch event details
     */
    private suspend fun fetchEventDetails(eventId: String) {
        try {
            println("📥 Fetching event: $eventId")
            // Supabase REST API requires 'eq.' prefix for equality filter
            val events = api.getEvent("eq.$eventId")
            println("📦 Events response: ${events.size} events")
            
            if (events.isNotEmpty()) {
                val event = events[0]
                println("🎫 Event: ${event.name}, Date: ${event.date}")
                _currentEvent.value = EventInfo(
                    id = event.id,
                    name = event.name,
                    date = event.date,
                    location = event.location?.get("address") as? String
                )
                println("✅ Event details saved")
            } else {
                println("⚠️ No event found for ID: $eventId")
            }
        } catch (e: Exception) {
            println("❌ Failed to fetch event: ${e.message}")
            e.printStackTrace()
            throw Exception("Failed to fetch event details: ${e.message}")
        }
    }
    
    /**
     * Start scanning
     */
    fun startScanning() {
        _isScanning.value = true
    }
    
    /**
     * Stop scanning
     */
    fun stopScanning() {
        _isScanning.value = false
    }
    
    /**
     * Validate ticket
     */
    suspend fun validateTicket(qrCode: String): ScanResult {
        try {
            val response = api.validateTicket(mapOf("qrCode" to qrCode))
            
            if (response.valid) {
                _scannedCount.value += 1
                
                // Record usage
                _scannerCodeId.value?.let { recordUsage(it) }
            }
            
            return ScanResult(
                success = response.valid,
                message = response.message ?: if (response.valid) "Ticket validated successfully" else "Invalid ticket",
                ticketInfo = response.ticket?.let {
                    TicketInfo(
                        holderName = it.holder_name,
                        holderEmail = it.holder_email,
                        ticketType = it.ticket_type
                    )
                }
            )
        } catch (e: Exception) {
            throw Exception("Validation failed: ${e.message}")
        }
    }
    
    /**
     * Record scanner usage
     */
    private suspend fun recordUsage(scannerCodeId: String) {
        try {
            api.recordScannerUsage(
                mapOf(
                    "p_scanner_code_id" to scannerCodeId,
                    "p_location" to null
                )
            )
        } catch (e: Exception) {
            // Silently fail, don't disrupt scanning
            e.printStackTrace()
        }
    }
    
    /**
     * Logout
     */
    fun logout() {
        _isAuthenticated.value = false
        _currentEvent.value = null
        _scannerCodeId.value = null
        scannerCode = null
        _scannedCount.value = 0
        _isScanning.value = false
        _sessionStartTime.value = null
        clearSession()
    }
}

File: mobile/ios/EventNexusScanner/ScannerViewModel.swift
//
//  ScannerViewModel.swift
//  EventNexusScanner
//
//  Main view model handling scanner state and API communication
//

import SwiftUI
import Combine

struct ScanResult {
    let success: Bool
    let message: String
    let ticketInfo: TicketInfo?
}

struct TicketInfo {
    let holderName: String
    let holderEmail: String
    let ticketType: String
}

struct EventInfo: Codable {
    let id: String
    let name: String
    let date: String
    let location: String?
}

class ScannerViewModel: ObservableObject {
    @Published var isAuthenticated = false
    @Published var currentEvent: EventInfo?
    @Published var scannerCodeId: String?
    @Published var scannedCount = 0
    @Published var isConnected = true
    @Published var sessionStartTime: Date?
    
    private var scannerCode: String?
    private var supabaseUrl = "https://anlivujgkjmajkcgbaxw.supabase.co"
    private var supabaseAnonKey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImFubGl2dWpna2ptYWprY2diYXh3Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjU5OTY0OTQsImV4cCI6MjA4MTU3MjQ5NH0.5SzkZg_PMqgdMClS1ftg4ZT_Ddyq1zOi-ZOLe1yuRgY"
    
    var sessionDuration: String {
        guard let startTime = sessionStartTime else { return "00:00" }
        let duration = Int(Date().timeIntervalSince(startTime))
        let hours = duration / 3600
        let minutes = (duration % 3600) / 60
        return hours > 0 ? String(format: "%d:%02d", hours, minutes) : String(format: "%02d:%02d", minutes, duration % 60)
    }
    
    // Authenticate with scanner code
    func authenticateWithCode(_ code: String) async throws {
        guard !code.isEmpty else {
            throw ScannerError.invalidCode
        }
        
        // Call verify_scanner_code RPC
        let url = URL(string: "\(supabaseUrl)/rest/v1/rpc/verify_scanner_code")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(supabaseAnonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(supabaseAnonKey)", forHTTPHeaderField: "Authorization")
        
        let body = ["p_code": code]
        request.httpBody = try JSONEncoder().encode(body)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ScannerError.networkError
        }
        
        guard httpResponse.statusCode == 200 else {
            throw ScannerError.authenticationFailed
        }
        
        // Parse response
        let results = try JSONDecoder().decode([VerifyCodeResponse].self, from: data)
        
        guard let result = results.first, result.valid else {
            throw ScannerError.invalidCode
        }
        
        // Store authentication details
        self.scannerCode = code
        self.scannerCodeId = result.scanner_code_id
        
        // Fetch event details
        if let eventId = result.event_id {
            try await fetchEventDetails(eventId: eventId)
        }
        
        await MainActor.run {
            self.isAuthenticated = true
            self.sessionStartTime = Date()
        }
        
        // Record scanner usage
        if let scannerCodeId = self.scannerCodeId {
            await recordUsage(scannerCodeId: scannerCodeId)
        }
    }
    
    // Fetch event details
    private func fetchEventDetails(eventId: String) async throws {
        let url = URL(string: "\(supabaseUrl)/rest/v1/events?id=eq.\(eventId)&select=*")!
        var request = URLRequest(url: url)
        request.setValue(supabaseAnonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(supabaseAnonKey)", forHTTPHeaderField: "Authorization")
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw ScannerError.networkError
        }
        
        let events = try JSONDecoder().decode([EventResponse].self, from: data)
        
        if let event = events.first {
            await MainActor.run {
                self.currentEvent = EventInfo(
                    id: event.id,
                    name: event.name,
                    date: event.date,
                    location: event.location?["address"] as? String
                )
            }
        }
    }
    
    // Validate ticket
    func validateTicket(qrCode: String) async throws -> ScanResult {
        guard let eventId = currentEvent?.id else {
            throw ScannerError.notAuthenticated
        }
        
        // Call validate-ticket Edge Function
        let url = URL(string: "\(supabaseUrl)/functions/v1/validate-ticket")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(supabaseAnonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(supabaseAnonKey)", forHTTPHeaderField: "Authorization")
        
        let body = ["qrCode": qrCode]
        request.httpBody = try JSONEncoder().encode(body)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ScannerError.networkError
        }
        
        guard httpResponse.statusCode == 200 else {
            throw ScannerError.validationFailed
        }
        
        let result = try JSONDecoder().decode(ValidationResponse.self, from: data)
        
        if result.valid {
            await MainActor.run {
                self.scannedCount += 1
            }
            
            // Record usage
            if let scannerCodeId = self.scannerCodeId {
                await recordUsage(scannerCodeId: scannerCodeId)
            }
        }
        
        return ScanResult(
            success: result.valid,
            message: result.message ?? (result.valid ? "Ticket validated successfully" : "Invalid ticket"),
            ticketInfo: result.ticket != nil ? TicketInfo(
                holderName: result.ticket?.holder_name ?? "Unknown",
                holderEmail: result.ticket?.holder_email ?? "",
                ticketType: result.ticket?.ticket_type ?? "General"
            ) : nil
        )
    }
    
    // Record scanner usage
    private func recordUsage(scannerCodeId: String) async {
        let url = URL(string: "\(supabaseUrl)/rest/v1/rpc/record_scanner_usage")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(supabaseAnonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(supabaseAnonKey)", forHTTPHeaderField: "Authorization")
        
        let body = ["p_scanner_code_id": scannerCodeId, "p_location": nil] as [String : Any?]
        if let jsonData = try? JSONSerialization.data(withJSONObject: body) {
            request.httpBody = jsonData
            
            do {
                _ = try await URLSession.shared.data(for: request)
            } catch {
                print("Failed to record usage: \(error)")
            }
        }
    }
    
    // Logout
    func logout() {
        isAuthenticated = false
        currentEvent = nil
        scannerCode = nil
        scannerCodeId = nil
        scannedCount = 0
        sessionStartTime = nil
    }
}

// MARK: - Response Models

struct VerifyCodeResponse: Codable {
    let valid: Bool
    let event_id: String?
    let event_name: String?
    let scanner_code_id: String?
    let organizer_id: String?
    let expires_at: String?
}

struct EventResponse: Codable {
    let id: String
    let name: String
    let date: String
    let location: [String: Any]?
    
    enum CodingKeys: String, CodingKey {
        case id, name, date, location
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        name = try container.decode(String.self, forKey: .name)
        date = try container.decode(String.self, forKey: .date)
        
        if let locationData = try? container.decode([String: AnyCodable].self, forKey: .location) {
            location = locationData.mapValues { $0.value }
        } else {
            location = nil
        }
    }
}

struct ValidationResponse: Codable {
    let valid: Bool
    let message: String?
    let ticket: TicketResponse?
}

struct TicketResponse: Codable {
    let holder_name: String
    let holder_email: String
    let ticket_type: String
}

// Helper for any JSON value
struct AnyCodable: Codable {
    let value: Any
    
    init(_ value: Any) {
        self.value = value
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        
        if let string = try? container.decode(String.self) {
            value = string
        } else if let int = try? container.decode(Int.self) {
            value = int
        } else if let double = try? container.decode(Double.self) {
            value = double
        } else if let bool = try? container.decode(Bool.self) {
            value = bool
        } else {
            value = ""
        }
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        
        if let string = value as? String {
            try container.encode(string)
        } else if let int = value as? Int {
            try container.encode(int)
        } else if let double = value as? Double {
            try container.encode(double)
        } else if let bool = value as? Bool {
            try container.encode(bool)
        }
    }
}

// MARK: - Errors

enum ScannerError: LocalizedError {
    case invalidCode
    case authenticationFailed
    case notAuthenticated
    case networkError
    case validationFailed
    
    var errorDescription: String? {
        switch self {
        case .invalidCode:
            return "Invalid scanner code. Please check and try again."
        case .authenticationFailed:
            return "Authentication failed. The scanner code may be expired or inactive."
        case .notAuthenticated:
            return "Please authenticate first."
        case .networkError:
            return "Network error. Please check your connection."
        case .validationFailed:
            return "Failed to validate ticket."
        }
    }
}
